### DISCUSSION TITLE
Creating features summarizing node degree info for Heterogeneous graphs

### DISCUSSION THREAD

**Discussion by fredjo89** [pyg-team/pytorch_geometric#3598] [Q&A] [answered]
Hello, 

I am currently working on adding more node features to my heterogeneous graph. I came accross the function LocalDegreeProfile() in the transforms-module, which I find really helpful! This function creates five more features for each node, that captures information of the nodes own degree and also neighboring node degree. 

The problem is that it doesen't seem to be implemented for heterogeneous graphs, which is why i'm wondering if anything similar exists for heterogeneous graphs, or any other advice on how do something similar with heterogeneous graph. 

To be specific: Using the example ('author', 'writes', 'paper'), for all nodes of type 'aurthor', I want to count the number of nodes 'paper' which it is linked to through the edge-type 'writes'

Thanks!

**Comment by wsad1** [comment, marked as answer]
Thanks for bringing this up. Currently there's nothing similar for heterogenous graphs. But it will be useful to add such support to `LocalDegreeProfile`. And the implementation would have to handle neighborhoods through different edge types.

You could use something similar to the below code to achieve this for now
```
N # number of author nodes
row_author, col_paper = data['author', 'writes', 'paper'].edge_index
row_paper, col_author = data['paper', 'written_by', 'author'].edge_index

 # degree of author for edge type ('author', 'writes', 'paper')
deg_author =  degree(row_author, N, dtype=torch.float)

# degree of paper nodes for edge type ('paper', 'written_by', 'author'), connected to author.
# OR how many authors does each paper node have.
deg = degree(row_paper, N, dtype=torch.float)
deg_col = deg[col_paper]

min_deg, _ = scatter_min(deg_col, row_author, dim_size=N)
min_deg[min_deg > 10000] = 0
max_deg, _ = scatter_max(deg_col, row_author, dim_size=N)
max_deg[max_deg < -10000] = 0
mean_deg = scatter_mean(deg_col, row_author, dim_size=N)
std_deg = scatter_std(deg_col, row_author, dim_size=N)
```

**Reply by rusty1s** [reply]
It's indeed interesting to bring this to heterogeneous graphs. One could do this in two separate ways:
1. Ignore edge types and treat the `LocalDegreeProfile` as homogeneous.
2. Apply the `LocalDegreeProfile` for each edge type separately, and concatenate the resulting features in the feature dimension.

**Reply by fredjo89** [reply]
Thank you so much for your help! I am indeed able to achieve what I want using the code provided by @wsad1 . 

FYI: In my attempt to utilize the information in my heterogeneous graph to a larger extent, I will proceed with approach #2 of @rusty1s suggestions. As for our previous example, number of citations and number of authors both sound like useful features for a paper, but adding them together might not make too much sense. 

Thank you for all your great work!

**Reply by rusty1s** [reply]
Please feel free to contribute if you make progress :)

**Comment by kunjanshah0811** [comment]
Is there any way to do this with ontolearn library?

**Reply by rusty1s** [reply]
No experience with this one. Can you explain what you want to do exactly?
